### DISCUSSION TITLE
How does the composable architecture play with the VisionOS ECS architecture?

### DISCUSSION THREAD

**Discussion by littlejohntj** [pointfreeco/swift-composable-architecture#2360] [General]
Starting to go deep into VisionOS development and I would love to use TCA inside of my app but I'm not sure how to have it play nicely with the Entity Component System that Apple forces you into. 

For example, if I want to have control within each frame, I need to initiate my logic within my System class. I can then use this context to query for my entities and then grab my components off of those. I could see making my Components reducers since you store state on these reducers but I'm not quite sure how I would access my store here. Especially since you don't even initialize the System object. 

Any thoughts or recommendations?

**Comment by mbrandonw** [comment]
Hi @littlejohntj, Stephen and I personally have not spent any time with VisionOS yet, though we hope to at some point. And hopefully the community can weigh in on this discussion too.

All I can say is that during this exploratory moment of seeing how to use TCA in VisionOS I highly caution against doing things such as introducing reference types to your TCA state, or sending actions with high frequency to a `Store` based on what is happening in the scene. Doing those things is just going to leave you in a bad spot.

In our [isowords](http://github.com/pointfreeco/isowords) project we do have TCA code that interacts with SceneKit, which is kinda like the predecessor to RealityKit, and that may give you some ideas how we interface between the two worlds.

**Reply by andreyz** [reply]
I second @mbrandonw in that isowords has a great example of real-life code integrating SceneKit and TCA. 

As I am writing a RealityKit based app myself, which is a blend of "vanilla" UI and 3D, I keep as much of RealityKit related code encapsulated as possible (ECS-related and other things like camera manipulation) and make sure that RealityKit code only reacts to changes in state and sends actions back to TCA. 

As @mbrandonw said, keep the state and actions concerning RealityKit as high level as possible (for example do not save any Entities or RealityKit classes in state) and do not send above-mentioned classes in actions. Instead focus on your app's domain model and do translation inside the RealityKit module.
